### PR TITLE
Add api check scripts

### PR DESCRIPTION
**Pull request by Chaoba**
PRs must be submitted under the terms of our Contributor License Agreement [CLA](https://github.com/mapbox/mapbox-maps-android/blob/main/CONTRIBUTING.md#contributor-license-agreement).
Fixes: < Link to related issues that will be fixed by this pull request, if they exist >

## Pull request checklist:
 - [ ] Briefly describe the changes in this PR.
 - [ ] Include before/after visuals or gifs if this PR includes visual changes.
 - [ ] Write tests for all new functionality. If tests were not written, please explain why.
 - [ ] Add example if relevant.
 - [ ] Document any changes to public APIs.
 - [ ] Apply changelog label ('breaking change', 'bug :beetle:', 'build', 'docs', 'feature :green_apple:', 'performance :zap:', 'testing :100:') or use the label 'skip changelog'
 - [ ] Add an entry inside this element for inclusion in the `mapbox-maps-android` changelog: `<changelog></changelog>`.

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -24,6 +24,24 @@ build:
...
+	sh ./scripts/java-api-check.sh "$(TAG)" "./plugin-lifecycle/build/outputs/aar/plugin-lifecycle-release.aar" "lifecycle"
+	sh ./scripts/java-api-check.sh "$(TAG)" "./plugin-gestures/build/outputs/aar/plugin-gestures-release.aar" "gestures"
+	sh ./scripts/java-api-check.sh "$(TAG)" "./plugin-compass/build/outputs/aar/plugin-compass-release.aar" "compass"
+	sh ./scripts/java-api-check.sh "$(TAG)" "./plugin-attribution/build/outputs/aar/plugin-attribution-release.aar" "attribution"
+	sh ./scripts/java-api-check.sh "$(TAG)" "./plugin-annotation/build/outputs/aar/plugin-annotation-release.aar" "annotation"
+	sh ./scripts/java-api-check.sh "$(TAG)" "./plugin-animation/build/outputs/aar/plugin-animation-release.aar" "animation"
+	sh ./scripts/java-api-check.sh "$(TAG)" "./extension-style/build/outputs/aar/extension-style-release.aar" "style"
+	sh ./scripts/java-api-check.sh "$(TAG)" "./extension-localization/build/outputs/aar/extension-localization-release.aar" "localization"
+	sh ./scripts/java-api-check.sh "$(TAG)" "./extension-androidauto/build/outputs/aar/extension-androidauto-release.aar" "androidauto"
+
```
Comment (tobrun) [inline Makefile:44]: I think we can rework this to be a for loop or embed it into a script or something like it to clean this a bit up.

Comment (Chaoba) [reply Makefile:44]: There is no clear rule in module name, aar name, and s3 link, so will move this part into a script.

File: .circleci/config.yml [hunk 2]
```diff
@@ -473,5 +486,7 @@ jobs:
           variant: "Release"
       - update-version-name
       - track-metrics
+      - init-aws
+      - check-api
```
Comment (tobrun) [inline .circleci/config.yml:490]: What is the reason/requirement for adding this to:  `build-sdk-release`?
Depending on how long it runs, I would opt to either add it to verify-code
or as a standalone `verify-api` (similar to verify-docs).

image

**Review by tobrun — COMMENTED** [review]
1 remark on when this is executed on CI env. , looks good otherwise

File: .circleci/config.yml (hunk 2 above)
Comment (Chaoba) [reply .circleci/config.yml:490]: I would prefer a standalone `verify-api`. Will add it later.

File: .circleci/config.yml [hunk 3]
```diff
@@ -331,6 +340,25 @@ jobs:
...
+  verify-api:
+    docker:
+      - image: mbgl/android-ndk-r21e:latest
+
+    steps:
+      - checkout
+      - restore-gradle-cache
+      - init-aws
+      - run:
+          no_output_timeout: 30m
```
Comment (tobrun) [inline .circleci/config.yml:352]: that is a long time :smile: :see_no_evil:

**Review by tobrun — APPROVED** [review]
:shipit: 
Curious to see how this will play out!

File: .circleci/config.yml (hunk 3 above)
Comment (Chaoba) [reply .circleci/config.yml:352]: In the CI, it actually cost 12m 26s 😄

File: .gitignore [hunk 4]
```diff
@@ -22,3 +22,4 @@ sdk/src/main/res/values/developer-config.xml
 node_modules/
 gh_token.txt
 android-docs-repo
+/scripts/revapi/revapi-0.9.7/
```
Comment (pengdev) [inline .gitignore:25]: Is this a temp folder? What if the version of the revapi is changed?

File: Makefile [hunk 5]
```diff
@@ -23,6 +25,11 @@ build:
 	./gradlew plugin-logo:assembleRelease;
 	./gradlew plugin-overlay:assembleRelease;
 	./gradlew plugin-scalebar:assembleRelease;
+	./gradlew plugin-lifecycle:assembleRelease;
+
+.PHONY: checkApi
+checkApi:
+	./scripts/java-api-check-all.sh "$(TAG)"
```
Comment (pengdev) [inline Makefile:32]: Would be good to add some documentation on what's the TAG stand for, and how to use it from command line interface.

File: scripts/java-api-check-all.sh [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+"${CURRENT_DIR}"/java-api-check.sh "$RELEATE_TAG" "./plugin-lifecycle/build/outputs/aar/plugin-lifecycle-release.aar" "lifecycle"
+"${CURRENT_DIR}"/java-api-check.sh "$RELEATE_TAG" "./plugin-gestures/build/outputs/aar/plugin-gestures-release.aar" "gestures"
+"${CURRENT_DIR}"/java-api-check.sh "$RELEATE_TAG" "./plugin-compass/build/outputs/aar/plugin-compass-release.aar" "compass"
+"${CURRENT_DIR}"/java-api-check.sh "$RELEATE_TAG" "./plugin-attribution/build/outputs/aar/plugin-attribution-release.aar" "attribution"
+"${CURRENT_DIR}"/java-api-check.sh "$RELEATE_TAG" "./plugin-annotation/build/outputs/aar/plugin-annotation-release.aar" "annotation"
+"${CURRENT_DIR}"/java-api-check.sh "$RELEATE_TAG" "./plugin-animation/build/outputs/aar/plugin-animation-release.aar" "animation"
+"${CURRENT_DIR}"/java-api-check.sh "$RELEATE_TAG" "./extension-style/build/outputs/aar/extension-style-release.aar" "style"
+"${CURRENT_DIR}"/java-api-check.sh "$RELEATE_TAG" "./extension-localization/build/outputs/aar/extension-localization-release.aar" "localization"
+"${CURRENT_DIR}"/java-api-check.sh "$RELEATE_TAG" "./extension-androidauto/build/outputs/aar/extension-androidauto-release.aar" "androidauto"
```
Comment (pengdev) [inline scripts/java-api-check-all.sh:40]: We can use similar scripts in https://github.com/mapbox/mobile-metrics/blob/8d86ab4d027bbbcff37664324f0be562b887acbe/android-maps-carbon/prepare_files.sh#L30 to get the list of modules from settings.gradle.kts and then filter modules start with `sdk-`, `plugin-`, `extension-`

**Review by pengdev — APPROVED** [review]
LGTM with a few nits

File: .gitignore (hunk 4 above)
Comment (Chaoba) [reply .gitignore:25]: This is not needed anymore, will delete it.

File: scripts/java-api-check-all.sh (hunk 6 above)
Comment (Chaoba) [reply scripts/java-api-check-all.sh:40]: Good point 👍

File: scripts/java-api-check-all.sh [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+    wget https://search.maven.org/remotecontent?filepath=org/revapi/revapi-standalone/0.9.7/revapi-standalone-0.9.7-standalone.zip -O revapi.zip \
+    && unzip revapi.zip \
+    && export PATH=${REVAPI}/revapi-0.9.7:$PATH
+    popd > /dev/null
+fi
+
+RELEATE_TAG=${1-""}
+echo "Release tag: $RELEATE_TAG"
+MODULES=$(cat ./settings.gradle.kts | grep include | sed 's/.*(\([^]]*\)).*/\1/g' | tr , \\n | sed 's/.*":\([^]]*\)".*/\1/g')
+PLUGINS=$(echo "$MODULES" | grep plugin-)
```
Comment (pengdev) [inline scripts/java-api-check-all.sh:27]: nit, PLUGINS is not used

File: scripts/java-api-check-all.sh [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+    MODULE_NAME=$i
+    if [[ $i == plugin* ]]; then
+      MODULE_NAME=${MODULE_NAME: 7}
+    elif [[ $i == extension* ]]; then
+      MODULE_NAME=${MODULE_NAME: 10}
+    elif [[ $i == *telemetry* ]]; then
+      continue
+    fi
+    echo "Checking module: $i, module aar name: $MODULE_NAME"
+    "${CURRENT_DIR}"/java-api-check.sh "$RELEATE_TAG" "./$i/build/outputs/aar/$i-release.aar" "$MODULE_NAME"
```
Comment (pengdev) [inline scripts/java-api-check-all.sh:39]: better to check if the release.aar file exists, we might want to skip app modules such as `extension-style-app` etc.

Comment (pengdev) [reply scripts/java-api-check-all.sh:39]: sorry, noticed we have checked the existence of .aar file already

**Pull request merged**